substrate/primitives/runtime-interface/src/pass_by.rs

```diff
@@ -15,419 +15,667 @@
+					type_name::<Self::FFIType>(),
+					type_name::<Self::Inner>()
+				);
+			},
+		}
+	}
 }
 
-/// The implementation of the pass by enum strategy. This strategy uses an `u8` internally to pass
-/// the enum between wasm and the host. So, this strategy only supports enums with unit variants.
+/// (DEPRECATED) Return `T` as a blob of bytes into the runtime.
 ///
-/// Use this type as associated type for [`PassBy`] to implement this strategy for a type.
+/// Uses `T::AsRef<[u8]>` to cast `T` into a `&[u8]`, allocates runtime memory
+/// using the legacy allocator, copies the slice into the runtime memory, and
+/// returns a pointer to it.
 ///
-/// This type expects the type that wants to implement this strategy as generic parameter. Besides
-/// that the type needs to implement `TryFrom<u8>` and `From<Self> for u8`.
+/// THIS STRATEGY IS DEPRECATED; DO NOT USE FOR NEW HOST FUNCTIONS!
 ///
-/// [`PassByEnum`](derive.PassByEnum.html) is a derive macro to implement this strategy.
-///
-/// # Example
-/// ```
-/// # use sp_runtime_interface::pass_by::{PassBy, Enum};
-/// #[derive(Clone, Copy)]
-/// enum Test {
-///     Test1,
-///     Test2,
-/// }
-///
-/// impl From<Test> for u8 {
-///     fn from(val: Test) -> u8 {
-///         match val {
-///             Test::Test1 => 0,
-///             Test::Test2 => 1,
-///         }
-///     }
-/// }
+/// Raw FFI type: `u32` (a pointer to the byte blob)
+pub struct AllocateAndReturnPointer<T, const N: usize>(PhantomData<(T, [u8; N])>);
+
+impl<T, const N: usize> RIType for AllocateAndReturnPointer<T, N> {
+	type FFIType = u32;
+	type Inner = T;
+}
+
+#[cfg(not(substrate_runtime))]
+impl<T, const N: usize> IntoFFIValue for AllocateAndReturnPointer<T, N>
+where
+	T: AsRef<[u8]>,
+{
+	fn into_ffi_value(value: Self::Inner, context: &mut dyn FunctionContext) -> Result<u32> {
+		let value = value.as_ref();
+		let addr = context.allocate_memory(value.len() as u32)?;
+		context.write_memory(addr, value)?;
+		Ok(addr.into())
+	}
+}
+
+#[cfg(substrate_runtime)]
+impl<T: codec::Decode, const N: usize> FromFFIValue for AllocateAndReturnPointer<T, N>
+where
+	T: From<[u8; N]>,
+{
+	fn from_ffi_value(arg: Self::FFIType) -> Self::Inner {
+		let value = unsafe { Vec::from_raw_parts(arg as *mut u8, N, N) };
+		let array = unsafe { *(value.as_ptr() as *const [u8; N]) };
+		T::from(array)
+	}
+}
+
+/// (DEPRECATED) Return `T` as a blob of bytes into the runtime.
 ///
-/// impl TryFrom<u8> for Test {
-///     type Error = ();
+/// Uses `T::AsRef<[u8]>` to cast `T` into a `&[u8]`, allocates runtime memory
+/// using the legacy allocator, copies the slice into the runtime memory, and
+/// returns a pointer to it.
 ///
-///     fn try_from(val: u8) -> Result<Test, ()> {
-///         match val {
-///             0 => Ok(Test::Test1),
-///             1 => Ok(Test::Test2),
-///             _ => Err(()),
-///         }
-///     }
-/// }
+/// THIS STRATEGY IS DEPRECATED; DO NOT USE FOR NEW HOST FUNCTIONS!
```

Comment:
Should we point towards what potential users should use instead of this?

Reply:
I added a short note.